File: backend/app/repositories/project.py
"""Project model."""

from datetime import datetime
from typing import List, Optional

from beanie import Document
from pydantic import Field


class ProjectMember(Document):
    """Project member schema."""

    user_id: str
    role: str = Field(pattern="^(owner|editor|viewer)$")
    joined_at: datetime = Field(default_factory=datetime.utcnow)


class Project(Document):
    """Project document model."""

    name: str = Field(..., min_length=3, max_length=50)
    subtitle: Optional[str] = Field(None, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    course_id: Optional[str] = Field(default=None)
    owner_id: str
    leader_id: Optional[str] = Field(default=None)
    members: List[dict] = Field(default_factory=list)  # List of ProjectMember-like dicts
    progress: int = Field(default=0, ge=0, le=100)
    is_template: bool = False
    is_archived: bool = False
    experiment_version: Optional[dict] = None
    inherited_template_key: Optional[str] = None
    inherited_template_label: Optional[str] = None
    inherited_template_release_id: Optional[str] = None
    inherited_template_source: Optional[str] = None
    initial_task_document_id: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        """Beanie settings."""

        name = "projects"
        indexes = [
            [("owner_id", 1)],
            [("leader_id", 1)],
            [("course_id", 1)],
            [("members.user_id", 1)],
            [("is_archived", 1)],
            [("created_at", 1)],  # For sorting projects by creation date
            [("owner_id", 1), ("is_archived", 1)],  # For user's active projects
            [("is_template", 1)],  # For template queries
            [("progress", 1)],  # For progress filtering
        ]

File: frontend/src/types/domain/project.ts
export interface Project {
    id: string
    name: string
    subtitle?: string
    description?: string
    course_id?: string
    owner_id: string
    leader_id?: string | null
    members: ProjectMember[]
    progress: number
    is_template: boolean
    is_archived: boolean
    inherited_template_key?: string
    inherited_template_label?: string
    inherited_template_release_id?: string
    inherited_template_source?: string
    initial_task_document_id?: string
    created_at: string
    updated_at: string
    experiment_version?: ExperimentVersion
}

export interface ProjectMember {
    user_id: string
    role: 'owner' | 'editor' | 'viewer'
    joined_at: string
}

export interface CreateProjectData {
    name: string
    subtitle?: string
    description?: string
    course_id?: string
    inherit_course_template?: boolean
    members?: string[]
    is_template?: boolean
}

export interface UpdateProjectData {
    name?: string
    subtitle?: string
    description?: string
    progress?: number
    is_archived?: boolean
    leader_id?: string | null
}

export interface ExperimentVersion {
    project_id: string
    mode: 'default' | 'research'
    version_name: string
    stage_control_mode: 'soft_guidance' | 'hard_constraint'
    process_scaffold_mode: 'on' | 'off'
    ai_scaffold_mode: 'single_agent' | 'multi_agent'
    broadcast_stage_updates: boolean
    group_condition?: string | null
    enabled_scaffold_layers: string[]
    enabled_scaffold_roles: string[]
    enabled_rule_set?: string | null
    export_profile?: string | null
    stage_sequence: string[]
    current_stage?: string | null
    template_key?: string | null
    template_label?: string | null
    template_release_id?: string | null
    template_release_note?: string | null
    template_source?: string | null
    graph_version?: string | null
    source_course_id?: string | null
    template_bound_at?: string | null
    updated_at?: string | null
}
